testcase/unittest/test_transaction_cache.py

```diff
@@ -0,0 +1,105 @@
+from loopchain.blockchain.transactions import TransactionVersioner
+from loopchain.blockchain.types import ExternalAddress
+from loopchain.blockchain.types import Hash32
+from loopchain.blockchain.types import Signature
+from loopchain.crypto.signature import Signer
+
+tx_versioner = TransactionVersioner()
+target_attrs = ("_cache_verify_hash", "_cache_verify_signature")
+
+
+class TestTxCache:
+    @pytest.fixture
+    def tx(self):
+        """Transaction fixture"""
+        tb = TransactionBuilder.new(version="0x3", type_=None, versioner=tx_versioner )
+        tb.step_limit = 1000000
+        tb.value = 100000
+        tb.signer = Signer.from_prikey(os.urandom(32))
+        tb.to_address = ExternalAddress(os.urandom(20))
+        tb.nid = 3
+        tb.nonce = random.randint(0, 100000)
+        tb.data = "test"
+        tb.data_type = "message"
+
+        transaction = tb.build()
+
+        return transaction
+
+    def verify(self, tv: TransactionVerifier, tx: Transaction):
+        tv.verify_hash(tx)
+        tv.verify_signature(tx)
+
+        return tx
+
+    def test_if_not_verified(self, tx):
+        """Check that Transaction has no attribute until verification func has been invoked"""
+        if "cache" in dir(tx):
+            raise AttributeError("Something wrong. Transaction has cached value when initialized.")
+
+        for attr in target_attrs:
+            with pytest.raises(AttributeError):
+                getattr(tx, attr)
+
+    @pytest.mark.parametrize("tv_exc_flag", [True, False])
+    def test_successfully_verified(self, tx, tv_exc_flag):
+        """Check that verification result been cached successfully"""
+        tv = TransactionVerifier.new(tx.version,  tx.type(), tx_versioner, raise_exceptions=tv_exc_flag)
+
+        verified_tx = self.verify(tv, tx)
+
+        for attr in target_attrs:
+            assert getattr(verified_tx, attr)
+
+    @pytest.mark.parametrize("raise_exception", [True, False])
+    @pytest.mark.parametrize("expected_exc, target_attr, monkey_value", [
+        (TransactionInvalidHashError, "hash", Hash32.new()),
+        (TransactionInvalidSignatureError, "signature", Signature.new()),
+    ])
+    def test_raised_exception_when_verified(self, tx, raise_exception, expected_exc, target_attr, monkey_value, monkeypatch):
+        """Check that exception successfully cached when raised exception in verification func"""
+        orig_value = getattr(tx, target_attr)
+
+        # Monkeypatch
+        object.__setattr__(tx, target_attr, monkey_value)
+        assert getattr(tx, target_attr) == monkey_value
+
+        # Fake exception
+        tv = TransactionVerifier.new(tx.version,  tx.type(), tx_versioner, raise_exceptions=raise_exception)
+        if raise_exception:
+            with pytest.raises(expected_exc):
+                getattr(tv, f"verify_{target_attr}")(tx)
+        else:
+            getattr(tv, f"verify_{target_attr}")(tx)
```

Comment:
Checking exceptions occurred is needed.

Reply:
I tried. Please check and resolve this if resolved.